packages/vm/src/runBlock.ts

```diff
@@ -407,6 +416,51 @@ async function applyBlock(this: VM, block: Block, opts: RunBlockOpts): Promise<A
   return blockResults
 }
 
+/**
+ * This method runs the logic of EIP 2935 (save blockhashes to state)
+ * It will put the `parentHash` of the block to the storage slot of `block.number - 1` of the history storage contract.
+ * This contract is used to retrieve BLOCKHASHes in EVM if EIP 2935 is activated.
+ * In case that the previous block of `block` is pre-EIP-2935 (so we are on the EIP 2935 fork block), additionally
+ * also add the currently available past blockhashes which are available by BLOCKHASH (so, the past 256 block hashes)
+ * @param this The VM to run on
+ * @param block The current block to save the parent block hash of
+ */
+export async function accumulateParentBlockHash(this: VM, block: Block) {
+  const historyAddress = Address.fromString(
+    bigIntToHex(this.common.param('vm', 'historyStorageAddress'))
+  )
+
+  if ((await this.stateManager.getAccount(historyAddress)) === undefined) {
+    await this.evm.journal.putAccount(historyAddress, new Account())
+  }
+
+  async function putBlockHash(vm: VM, hash: Uint8Array, number: bigint) {
+    const key = setLengthLeft(bigIntToBytes(number), 32)
+    await vm.stateManager.putContractStorage(historyAddress, key, hash)
+  }
+  await putBlockHash(this, block.header.parentHash, block.header.number - BIGINT_1)
+
+  // Check if we are on the fork block
+  const forkTime = this.common.eipTimestamp(2935)
+  if (forkTime === null) {
+    throw new Error('EIP 2935 should be activated by timestamp')
+  }
+  const parentBlock = await this.blockchain.getBlock(block.header.parentHash)
+
+  if (parentBlock.header.timestamp < forkTime) {
+    let ancestor = parentBlock
+    const range = this.common.param('vm', 'minHistoryServeWindow')
+    for (let i = 0; i < Number(range) - 1; i++) {
```

Comment:
I would cautiously think that this `<` usage together with ` - 1` might produce an "off by one" error?

Since the code from the EIP states `for i in range(MIN_HISTORY_SERVE_WINDOW - 1)`?

(maybe confirm with edge case tests?)

Reply:
I think this is correct (but this is definitely off-by-one error prone! :smile: )

From EIP: `for i in range(MIN_HISTORY_SERVE_WINDOW - 1):` this is python code. This loops the for loop `MIN_HISTORY_SERVE_WINDOW - 1` time. Note that `MIN_HISTORY_SERVE_WINDOW` is 256 so we loop this 255 times. In the current TS code we loop `i < 256 - 1` = 255 times also. Note that for each loop, we store the ancestor block. Also note that previously we have already stored the parent block hash of the current block. So, in total we have stored 256 blocks and therefore the history contract has access to 256 slots.

Reply:
👍